cpp/src/search/search.cu

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iterator/iterator.cuh>
+#include <table/device_table.cuh>
+#include <table/device_table_row_operators.cuh>
+#include <utilities/wrapper_types.hpp>
+#include <utilities/column_utils.hpp>
+
+#include <cudf/search.hpp>
+#include <cudf/copying.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+
+#include <thrust/binary_search.h>
+#include <thrust/device_vector.h>
+
+namespace cudf {
+
+namespace {
+
+template <typename DataIterator, typename ValuesIterator, typename Comparator>
+void launch_search(DataIterator it_data,
+                    ValuesIterator it_vals,
+                    gdf_size_type data_size,
+                    gdf_size_type values_size,
+                    void* output,
+                    Comparator comp,
+                    bool find_first,
+                    cudaStream_t stream)
+{
+  if (find_first) {
+    thrust::lower_bound(rmm::exec_policy(stream)->on(stream),
+                        it_data, it_data + data_size,
+                        it_vals, it_vals + values_size,
+                        static_cast<gdf_index_type*>(output),
+                        comp);
+  }
+  else {
+    thrust::upper_bound(rmm::exec_policy(stream)->on(stream),
+                        it_data, it_data + data_size,
+                        it_vals, it_vals + values_size,
+                        static_cast<gdf_index_type*>(output),
+                        comp);
+  }
+}
+
+struct search_functor {
+  template <typename T>
+  void operator()(gdf_column const& column,
+                  gdf_column const& values,
+                  bool find_first,
+                  bool ascending,
+                  bool nulls_as_largest,
+                  cudaStream_t stream,
+                  gdf_column& result)
+  {
+    if ( is_nullable(column) ) {
+      T null_substitute = (nulls_as_largest) 
```

Comment:
This will be problematic if `std::numeric_limits<T>::max()` or `std::numeric_limits<T>::lowest()` values are present in the column you are searching. I suggest just using the `row_inequality_operator` even for the single column case because it will safely handle comparing nulls.